cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,315 @@
+      j]MJ]  jM]1]P .]M1 jMMP     MM]1  I        J  t   1   j J    '       
+      =M]dT  jM]q0' dM]  M]MT     ]MM  j        j   j  j    J 1            
+      ]M]M`  j]0j#  ]MF  ]M]'    .M]P  J       .'   j  J  .J  4_,          
+      M]0M   =MM]1 .M]'  MM]     jM](  1       r    j  1  --,   "!         
+      ]0MT   ]M]M  jM@   ]M]     M]P  j       J     j j     4     1        
+      MM]'   M]0P  j]1  .M]M    j]M'  J      j'     ",?     j     1        
+     _]M]    M]0`  jM1 .MNMM,  .]M'   1     .'       11     1    j'        
+     jM]1   jM]@   j]L_]'?M]M__MP'    \     J        1G    J    .'         
+     j]0(   jM]1   "M]P'  "N]M/-      "L__J L________'?L__- *__,'          
+     "-'    "--                                                            
+                                                                           
+----------------------------------------------------------------------------
+
+  ******************************************************************************
+  * @file    dma.c
+  * @author  MCD Application Team, Wi6labs
+  * @brief   dma service.
+  ******************************************************************************
+
+Copyright (c) <2017>, <Wi6labs>
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of the wi6labs nor the
+      names of its contributors may be used to endorse or promote products
+      derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL WI6LABS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+describe file content here
+-----------------------------------------------------------------------------*/
+#include "stm32_def.h"
+#include "dma.h"
+#include "analog.h"
+
+#ifdef __cplusplus
+ extern "C" {
+#endif
+
+extern ADC_HandleTypeDef AdcHandleTab[ADC_NUM];
+
```

Comment:
dma.c file should not have any reference to ADC, so that it can later be re-used by oher devices

Reply:
We will see if it possible in next release.

Reply:
After study, it's not possible to remove this variable from DMA. The DMA link a peripheral, it need to have the information that is in the ADC structure. If we don't share this variable between ADC and DMA, DMA will not work.

Reply:
The idea was to do the DMA allocation in dma.c and keep the dma actual configuration in each driver ADC and others later. Otherwise this file here will become huge and needs to be updated each time when you need to add DMA support for a new driver. If we make this file simpler, then up to each driver to add the calls to DMA based on the "allocated" DMA information.  My point is not that the code is wrong, but rather the split between the files can be changed

Reply:
If i understand well, dma_allocate() just give you the irqHandler and the channel to use, then all the initialization of the DMA will be done in analog.c file, in HAL_ADC_MspInit() function.

Reply:
That would be the idea of the proposal  indeed - please think about it and let let me know if you think it makes sense.  Also I think that the DMA mapping table can be generic and could contain any kind of peripheral instances (ADC, serial, I2C, etc ... ) so the name and types should be generic
Depending on peripherals that are used, the DMA settings will differ (periph to mem, or mem to periph, etc.) so it is rather up to the peripheral driver to set these options.

Reply:
Done, let me know if it's OK. I didn't rebase the git history yet if you want to go back.